### ISSUE TITLE
CI refactoring to improve reliability, maintainability, deployment speed

### ISSUE THREAD

**Issue by jamilbk** [CloudFire-LLC/cloudfire#3995]
## Goals

- Prevent production and published artifacts from drifting to incompatible versions
- Improve speed (Windows build...)
- Reduce duplicated config
- Develop intuition for backwards-incompatible changes

```[tasklist]
- [ ] #3321
- [x] When running CD, don't re-run tauri build in CI
- [x] `/actions/setup-elixir.yml`
- [x] Refactor hotfix into its own workflow and add a sanity check before deploying to terraform -- firezone/product#684
- [x] Add backwards compatibility matrix to integration-tests
- [x] Build release artifacts while Terraform is working (would reduce CD from 32.5 to 26 minutes)
- [x] Fix `cargo-chef` (#4033) or remove it (#4031)
```

**Comment by ReactorScram** [comment]
CI's bottleneck seems to be the Docker Client image https://github.com/firezone/firezone/pull/4031
CD's bottleneck is the Terraform `deploy-staging` step. All the builds wait on it. If we could kick off the builds when Terraform starts (when CI passes) and wait until Terraform passes before publishing the builds, that would save 6-7 minutes out of the 33-minute CD wall time

**Comment by ReactorScram** [comment]
~~Hm doesn't look any faster after bumping Rust for cargo-chef~~

<img width="989" alt="image" src="https://github.com/firezone/firezone/assets/13400041/d70278fb-4153-4839-9b75-bbacb5440745">

nvm it did work on this one, maybe it cache hits if the Rust code doesn't change? 
<img width="514" alt="image" src="https://github.com/firezone/firezone/assets/13400041/aea463a7-e565-4da6-81b4-1579dd572aa8">

**Comment by thomaseizinger** [comment]
> Improve speed (Windows build...)

The windows CI steps don't seem to perform any caching at the moment: https://github.com/firezone/firezone/actions/runs/8228842619/job/22498988922?pr=4069#step:5:29

**Comment by jamilbk** [comment]
@thomaseizinger ~~Yeah in the PR linked to this issue I've reverted to sccache. I don't think the Swatinem GH cache was much of a speedup, if any.~~ Edit: It shaves off about a minute when it's used, and we could probably shave it down further, so I've kept it.

Most of the goal of this issue is refactoring to parallelize jobs better and not run the tauri build twice, which it achieves.

We should split off another issue to look deeper into caching that build better.
